include/graph_contracts.hpp

```diff
@@ -0,0 +1,444 @@
+    { t.name() } -> std::same_as<std::string_view>;
+    { t.work() } -> std::same_as<void>; // primary function to be called by scheduler
+    { t.input_ports() } -> std::same_as<std::vector<dyn_port>&>;
+    { t.output_ports() } -> std::same_as<std::vector<dyn_port>&>;
+    //{ t.connect(src_port, dst_block, dst_port) } -> std::same_as<std::vector<connection_result_t>>;
+};
+
+template<typename T, fixed_string PortName, port_type_t PortType, port_direction_t PortDirection, // TODO: sort default arguments
+    std::size_t N_HISTORY = std::dynamic_extent, std::size_t MIN_SAMPLES = std::dynamic_extent, std::size_t MAX_SAMPLES = std::dynamic_extent, bool OPTIONAL = false,
+    gr::Buffer BufferType = gr::circular_buffer<T>>
+class port {
+    static_assert(PortDirection != port_direction_t::ANY, "ANY reserved for queries and not port direction declarations");
+public:
+    using value_type = T;
+    static constexpr bool IS_INPUT = PortDirection == port_direction_t::INPUT;
+private:
+    using ReaderType = decltype(std::declval<BufferType>().new_reader());
+    using WriterType = decltype(std::declval<BufferType>().new_writer());
+    using IoType = std::conditional_t<IS_INPUT, ReaderType, WriterType>;
+
+    const std::string _port_name;
+    const std::int16_t _priority = 0; // → dependents of a higher-prio port should be scheduled first (Q: make this by order of ports?)
+    const std::size_t _n_history = N_HISTORY;
+    const std::size_t _min_samples = MIN_SAMPLES;
+    const std::size_t _max_samples = MAX_SAMPLES;
+    bool _connected = false;
+
+    IoType _ioHandler = new_io_handler();
+
+    constexpr auto new_io_handler() noexcept {
+        if constexpr (IS_INPUT) { return BufferType(4096).new_reader();
+        } else { return BufferType(4096).new_writer(); }
+    }
+
+    void* writer_handler_internal() noexcept {
+        if constexpr (!IS_INPUT) {
+            static_assert(!IS_INPUT, "only to be used with output ports");
+        }
+        return static_cast<void*>(std::addressof(_ioHandler));
+    }
+    bool update_reader_internal(void* buffer_writer_handler_other) noexcept {
+        if constexpr (IS_INPUT) {
+            static_assert(IS_INPUT, "only to be used with input ports");
+        }
+        try {
+            auto typed_buffer_writer = static_cast<WriterType*>(buffer_writer_handler_other);
+            if (typed_buffer_writer == nullptr) {
+                return false;
+            }
+            setBuffer(typed_buffer_writer->get_buffer());
+        } catch (...) {
+            // needed for runtime polymorphism
+            assert(false && "invalid static_cast from 'void* -> WriterType*' in update_reader_internal()");
+            return false;
+        }
+        return true;
+    }
+    friend dyn_port;
+
+    public:
+        constexpr port() = default;
+        constexpr port(std::string_view port_name, std::int16_t priority = 0, std::size_t n_history = 0,
```

Comment:
It is a pessimization to convert to string_view  and then back to std::string when one passes a rvalue std::string for port name.

Receive by value, and std::move into the member.

Reply:
done